include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,372 @@
+
+                    auto __group = __item.get_group();
+                    ::std::size_t __global_id = __item.get_global_id();
+                    ::std::size_t __local_id = __group.get_local_id();
+
+                    // 1a. Perform a serial scan within the work item over assigned elements. Store partial
+                    // reductions in work item memory, and write the accumulated value and number of counted
+                    // segments into work group memory.
+                    ::std::size_t __start = __global_id * __vals_per_item;
+                    ::std::size_t __end = __dpl_sycl::__minimum<decltype(__n)>{}(__start + __vals_per_item, __n);
+
+                    // First work item must set their accumulator to the provided init
+                    if (__global_id == 0 && __local_id == 0)
+                    {
+                        __accumulator = __init;
+                    }
+
+                    const int64_t NO_SEGMENT_BREAK = -1;
+                    // signed to allow flag for no segment break found
+                    int64_t __max_end = NO_SEGMENT_BREAK;
+
+                    for (::std::size_t __i = __start; __i < __end; ++__i)
+                    {
+                        if constexpr (__scan_type == scan_type::inclusive)
+                        {
+                            __accumulator = __binary_op(__accumulator, __values[__i]);
+
+                            __out_values[__i] = __accumulator;
+                        }
+                        else // exclusive scan
+                        {
+                            __out_values[__i] = __accumulator;
+                            __accumulator = __binary_op(__accumulator, __values[__i]);
+                        }
+
+                        // reset the accumulator to init if we reach the end of a segment
+                        // (init for inclusive scan is the identity)
+                        if (__n - 1 == __i || !__binary_pred(__keys[__i], __keys[__i + 1]))
+                        {
+                            __accumulator = __init;
+                            __max_end = __local_id;
+                        }
+                    }
+
+                    // 1b. Perform a work group scan to find the carry in value to apply to each item.
+                    ::std::int64_t __closest_seg_id = __dpl_sycl::__inclusive_scan_over_group(
+                        __group, __max_end, __dpl_sycl::__maximum<decltype(__max_end)>());
+
+                    __flag_type __group_has_segment_break = (__closest_seg_id != NO_SEGMENT_BREAK);
+
+                    //get rid of no segment end found flag
+                    __closest_seg_id = ::std::max(::std::int64_t(0), __closest_seg_id);
+                    __val_type __carry_in = wg_segmented_scan(__item, __loc_acc, __local_id,
+                                                              __local_id - __closest_seg_id, __accumulator, __binary_op,
```

Comment:
Should you cast `__local_id` to a signed type? Otherwise, would there be a risk of integer underflow here?

Reply:
It is guaranteed that `__local_id >= __closest_seg_id` here since the search for the closest segment end is performed on lower indexed work items. However, it is possible that `__closest_seg_id` is -1. Is there any risk in subtracting -1 from an unsigned integer since the result is equivalent to adding one?